src/NationalInstruments.Analyzers/Correctness/RecordWithEnumerablesShouldOverrideDefaultEqualityAnalyzer.cs

```diff
@@ -0,0 +1,96 @@
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using NationalInstruments.Analyzers.Properties;
+using NationalInstruments.Analyzers.Utilities;
+using NationalInstruments.Analyzers.Utilities.Extensions;
+
+namespace NationalInstruments.Analyzers.Correctness
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public sealed class RecordWithEnumerablesShouldOverrideDefaultEqualityAnalyzer
+        : NIDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "NI1019";
+
+        public static readonly DiagnosticDescriptor Rule = new DiagnosticDescriptor(
+            DiagnosticId,
+            new LocalizableResourceString(nameof(Resources.NI1019_Title), Resources.ResourceManager, typeof(Resources)),
+            new LocalizableResourceString(nameof(Resources.NI1019_Message), Resources.ResourceManager, typeof(Resources)),
+            Category.Correctness,
+            DiagnosticSeverity.Warning,
+            isEnabledByDefault: true,
+            description: new LocalizableResourceString(nameof(Resources.NI1019_Description), Resources.ResourceManager, typeof(Resources)));
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecutionIf(IsRunningInProduction);
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.RegisterSymbolAction(AnalyzeRecord, SymbolKind.NamedType);
+        }
+
+        private void AnalyzeRecord(SymbolAnalysisContext context)
+        {
+            var typeSymbol = (ITypeSymbol)context.Symbol;
+            if (typeSymbol.IsRecord == false)
+            {
+                return;
+            }
+
+            var baseTypeProperties = GetBaseTypeProperties(typeSymbol)
+                .Select(p => p.Name)
+                .ToImmutableHashSet();
+
+            var enumerableProperties = typeSymbol
+                .GetPublicPropertySymbols()
+                .Where(p => p.Type.IsEnumerable() && !baseTypeProperties.Contains(p.Name))
```

Comment:
What's the point of ignoring the enumerable if the base type has a property with the same name? Doesn't that allow the following to _not_ report a diagnostic?

```cs
public record BaseRecord
{
    public int MyInts { get; }
}

public record TestRecord : BaseRecord
{
    public IEnumerable<int> MyInts { get; }
}
```

Reply:
My original intention was to make sure this case does not report a diagnostic:

```cs
public record BaseRecord(IEnumerable<int> MyInts)
{
    bool virtual Equals(BaseRecord? other)
    {
        // ...
    }

    int override GetHashCode()
    {
        //...
    }
}

public record TestRecord(IEnumerable<int> MyInts)
    : BaseRecord(MyInts);
```

Here, the base record defines `MyInts`, and does correctly implement `Equals`.  The `TestRecord` derives from it, and though it does has an enumerable property when inspected, the property is derived from the base class and is not defined by `TestRecord`—so no diagnostic should be reported even if `TestRecord` does not have an `Equals` implementation.

But you're right, the code you wrote above will cause the analyzer to incorrectly suppress the diagnostic.  Let me add your code as a test case, I think we can fix it by just adding an additional check to verify the `baseTypeProperty` type is the same as the derived record's property type.

Reply:
Updated the code to check the property type as well as the name.  Added tests to cover this case as well.

Reply:
Did you push those changes?

What if `TestRecord` in your example above was defined as `public record TestRecord(IEnumerable<int> OtherInts) : BaseRecord(OtherInts)`? Users will have to implement `Equals` for `TestRecord` because the names don't match?

Reply:
> Did you push those changes?

Yes.  It starts on line 44 in `RecordWithEnumerablesShouldOverrideDefaultEqualityAnalyzer.cs`:

```cs
var baseTypeProperties = GetBaseTypeProperties(typeSymbol)
    .ToImmutableHashSet(SymbolEqualityComparer.Default);

var enumerableProperties = typeSymbol
    .GetPublicPropertySymbols()
    .Where(p => p.Type.IsEnumerable() && !baseTypeProperties.Contains(p))
    .ToImmutableArray();
```

Instead of just tracking the property name, I'm tracking the property symbol itself (and using the symbol equality comparer, otherwise Contains() uses reference equals).

> What if TestRecord in your example above was defined as public record TestRecord(IEnumerable<int> OtherInts) : BaseRecord(OtherInts)? Users will have to implement Equals for TestRecord because the names don't match?

Yes that's correct.  If the derived record declares a differently named property, the compiler will generate a new property in the derived record,  so equality must be implemented.

You can see how the compiler generates the record in [sharplab.io](https://sharplab.io/#v2:CYLg1APgAgDABFAjAFgNwFgBQWoGY4BOApgMYD2BwCATHAEICGAzkQBRK4A8AlgHYAuAPjgBZAJ4BJAUwCUGbJjyFSFKlFoARIgW4A3IsHaIufIaMnSZWODbgh6zNuKn9Z8nPmLlKNOFp36wAAKBGQADgBKRLwMALYGRiYCwgDy/AAW2i6y1rb2jCysaZkE2XJAA).  I've posted the relevant code here: 

### Original user code

```cs
public record Base(IEnumerable<int> MyInts);

public record DerivedPropRenamed(IEnumerable<int> OtherInts)
    : Base(OtherInts);
```

### Generated base record

```cs
public class Base : IEquatable<Base>
{
    [CompilerGenerated]
    [DebuggerBrowsable(DebuggerBrowsableState.Never)]
    private readonly IEnumerable<int> <MyInts>k__BackingField;

    public IEnumerable<int> MyInts
    {
        [CompilerGenerated]
        get
        {
            return <MyInts>k__BackingField;
        }
        [CompilerGenerated]
        init
        {
            <MyInts>k__BackingField = value;
        }
    }

    public Base(IEnumerable<int> MyInts)
    {
        <MyInts>k__BackingField = MyInts;
        base..ctor();
    }
}
```
### Generated derived record with differently named property:

```cs
public class DerivedPropRenamed : Base, IEquatable<DerivedPropRenamed>
{
    [CompilerGenerated]
    [DebuggerBrowsable(DebuggerBrowsableState.Never)]
    private readonly IEnumerable<int> <OtherInts>k__BackingField;

    public IEnumerable<int> OtherInts
    {
        [CompilerGenerated]
        get
        {
            return <OtherInts>k__BackingField;
        }
        [CompilerGenerated]
        init
        {
            <OtherInts>k__BackingField = value;
        }
    }

    public DerivedPropRenamed(IEnumerable<int> OtherInts)
    {
        <OtherInts>k__BackingField = OtherInts;
        base..ctor(OtherInts);
    }
}
```

For completeness, I'll add this as a test case.

Reply:
Test case added.

Reply:
Cool, never heard of sharplab.io